Objects/genericaliasobject.c

```diff
@@ -416,21 +429,41 @@ _Py_subs_parameters(PyObject *self, PyObject *args, PyObject *parameters, PyObje
         t = list[T];          t[int]      -> newargs = [int]
         t = dict[str, T];     t[int]      -> newargs = [str, int]
         t = dict[T, list[S]]; t[str, int] -> newargs = [str, list[int]]
+        t = list[[T]];        t[str]      -> newargs = [[str]]
      */
-    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
+    Py_ssize_t nargs = PySequence_Length(args);
     PyObject *newargs = PyTuple_New(nargs);
     if (newargs == NULL) {
         Py_DECREF(item);
         return NULL;
     }
     for (Py_ssize_t iarg = 0, jarg = 0; iarg < nargs; iarg++) {
-        PyObject *arg = PyTuple_GET_ITEM(args, iarg);
+        PyObject *arg = PySequence_GetItem(args, iarg);
         if (PyType_Check(arg)) {
             PyTuple_SET_ITEM(newargs, jarg, Py_NewRef(arg));
             jarg++;
             continue;
         }
-
+        // Recursively substitute params in lists/tuples.
+        if (PyTuple_Check(arg) || PyList_Check(arg)) {
+            PyObject *subargs = _Py_subs_parameters(self, arg, parameters, item);
+            if (subargs == NULL) {
+                Py_DECREF(newargs);
+                Py_DECREF(item);
+                return NULL;
+            }
+            PyObject *subargs_list = PySequence_List(subargs);
+            Py_DECREF(subargs);
+            if (subargs_list == NULL) {
+                Py_DECREF(newargs);
+                Py_DECREF(item);
+                return NULL;
+            }
+            Py_SETREF(subargs, subargs_list);
+            PyTuple_SET_ITEM(newargs, jarg, Py_NewRef(subargs));
```

Comment:
I'm not entirely sure if the call to `Py_NewRef` is needed here?

Reply:
To validate this, try running tests in refleak mode locally, something like `./python.exe -m test -R : test_genericalias`.

Reply:
Thanks for the tip! I did have some leaks which I managed to fix :)